Wrappers/Python/cil/optimisation/utilities/preconditioner.py

```diff
@@ -0,0 +1,179 @@
+    
+        gradient : DataContainer
+            The calculated gradient to modify
+        out : DataContainer, optional, default is None 
+            The modified gradient to return 
+            
+        """
+        pass
+
+
+class Sensitivity(Preconditioner):
+
+    r"""
+    Sensitivity preconditioner class. 
+
+    In each call to the preconditioner the `gradient` is multiplied by :math:` y/(A^T \mathbf{1})` where :math:`A` is an operator, :math:`\mathbf{1}` is an object in the range of the operator filled with ones and :math:`y` is an optional reference image.  
+
+    Parameters
+    ----------
+    operator : CIL Operator 
+        The operator used for sensitivity computation.
+    reference : DataContainer e.g. ImageData, optional
+        The reference data, an object in the domain of the operator. Recommended to be a best guess reconstruction. 
+    """
+
+    def __init__(self, operator, reference=None):
+
+        super(Sensitivity, self).__init__()
+        self.operator = operator
+        self.reference = reference
+
+        self.compute_preconditioner_matrix()
+
+    def compute_preconditioner_matrix(self):
+        r"""
+        Compute the sensitivity. :math:`A^T \mathbf{1}` where :math:`A` is the operator and :math:`\mathbf{1}` is an object in the range of the operator filled with ones.        
+        Then perform safe division by the sensitivity to store the preconditioner array :math:` y/(A^T \mathbf{1})`
+        """
+        self.array = self.operator.adjoint(
+            self.operator.range_geometry().allocate(value=1.0))
+
+        if self.reference is not None:
+            self.reference.divide(
+                self.array, where=self.array.as_array() > 0, out=self.array)
+        else:
+            self.operator.range_geometry().allocate(value=1.0).divide(
+                self.array, where=self.array.as_array() > 0, out=self.array)
+
+    def apply(self, algorithm, gradient, out=None):
+        r"""
+        Update the preconditioner.
+
+        Parameters
+        ----------
+        algorithm : object
+            The algorithm object.
+        gradient : DataContainer
+            The calculated gradient to modify
+        out : DataContainer, optional, default is None 
+            The modified gradient to return 
+        """
+        if out is None:
+            out = gradient.copy()
+        gradient.multiply(
+            self.array, out=out)
+        return out
+
+
+class AdaptiveSensitivity(Sensitivity):
+
+    r"""
+    Adaptive Sensitivity preconditioner class. 
+
+    In each call to the preconditioner the `gradient` is multiplied by :math:` (x+\delta) \odot y /(A^T \mathbf{1})` where :math:`A` is an operator,  :math:`\mathbf{1}` is an object in the range of the operator filled with ones and :math:`y` is an optional reference image.  
+    The point :math:`x` is the current iteration and :math:`delta` is a small positive float. 
+
+
+    Parameters
+    ----------
+    operator : CIL object
```

Comment:
Have three options for x in the precondition  :
1. Fixed
2. Updates then freezes
3. Always updating